### PR TITLE
Docs for Collaborative Training

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds documentation for Collaborative Training!

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @SeanNaren

### CODE REVIEW

File: docs/source/strategies/collaborative_training_intermediate.rst [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+
+Re-using Gradient Buffers
+"""""""""""""""""""""""""
+
+By default, Hivemind accumulates gradients in a separate buffer. This means additional GPU memory is required to store gradients. You can disable this by passing `reuse_grad_buffers=False` to the `CollaborativeStrategy`.
+
+.. warning::
+    The `CollaborativeStrategy` will override `zero_grad` in your `LightningModule` to have no effect. This is because gradients are accumulated in the model
+    and hivemind manages when they need to be cleared.
```
Comment (rohitgr7) [inline docs/source/strategies/collaborative_training_intermediate.rst:88]:
```suggestion
By default, Hivemind accumulates gradients in a separate buffer. This means additional GPU memory is required to store gradients. You can disable this by passing ``reuse_grad_buffers=False`` to the ``CollaborativeStrategy``.

.. warning::
    The ``CollaborativeStrategy`` will override ``zero_grad`` in your ``LightningModule`` to have no effect. This is because gradients are accumulated in the model
    and hivemind manages when they need to be cleared.
```

File: docs/source/strategies/collaborative_training_intermediate.rst [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+By default, Hivemind accumulates gradients in a separate buffer. This means additional GPU memory is required to store gradients. You can disable this by passing ``reuse_grad_buffers=False`` to the ``CollaborativeStrategy``.
+
+.. warning::
+    The ``CollaborativeStrategy`` will override ``zero_grad`` in your ``LightningModule`` to have no effect. This is because gradients are accumulated in the model
+    and hivemind manages when they need to be cleared.
+
+.. code-block:: python
+
+    import torch
+    from functools import partial
```
Comment (kaushikb11) [inline docs/source/strategies/collaborative_training_intermediate.rst:93]: Is the partial import required?

File: docs/source/strategies/collaborative_training_basic.rst [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    pip install hivemind
+
+The ``CollaborativeStrategy`` accumulates gradients from all processes that are collaborating till they reach a ``target_batch_size``. By default, we use the batch size
+of the first batch to determine what each local machine batch contributes towards the ``target_batch_size``. Once the ``target_batch_size`` is reached, an optimizer step
+is made on all processes.
+
+.. warning::
+
+    When using ``CollaborativeStrategy`` note that you cannot use gradient accumulation (``accumulate_grad_batches``). This is because hivemind manages accumulation internally.
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_basic.rst:25]: Accumulation is not supported (it kinda is through hivemind using the global batch size which I think is kinda explanatory). I just realised I think I'll need another PR on top of this to enforce this check, will follow up!

File: docs/source/strategies/collaborative_training_intermediate.rst (hunk 2 above)
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_intermediate.rst:93]: Yep for the scheduler

File: docs/source/strategies/collaborative_training_basic.rst [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+Under the hood, we use `Hivemind <https://github.com/learning-at-home/hivemind>`_ which provides de-centralized training across the internet.
+
+To use Collaborative Training, you need to first install Hivemind.
+
+.. code-block:: bash
+
+    pip install hivemind
+
+The ``CollaborativeStrategy`` accumulates gradients from all processes that are collaborating till they reach a ``target_batch_size``. By default, we use the batch size
```
Comment (RobertLaurella) [inline docs/source/strategies/collaborative_training_basic.rst:19]:
```suggestion
The ``CollaborativeStrategy`` accumulates gradients from all processes that are collaborating until they reach a ``target_batch_size``. By default, we use the batch size
```

File: docs/source/strategies/collaborative_training_basic.rst [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+.. code-block:: bash
+
+    python train.py
+    # Other machines can connect running the same command:
+    # INITIAL_PEERS=... python train.py
+    # or passing the peers to the strategy:"
+    # CollaborativeStrategy(initial_peers=...)"
+
+
+Once training starts a helper message is printed, showing how to start training on other machines using the same code.
```
Comment (RobertLaurella) [inline docs/source/strategies/collaborative_training_basic.rst:43]:
```suggestion
A helper message is printed once your training begins, which shows you how to start training on other machines using the same code.
```

File: docs/source/strategies/collaborative_training_expert.rst [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+.. _collaborative_training_expert:
+
+Training on unreliable mixed GPUs across the internet (Expert)
+==============================================================
+
+Using Compression to Optimize communications
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Below are some ways to reduce communication when training collaboratively. As the model sizes get larger, bottlenecks in communication become more apparent.
```
Comment (RobertLaurella) [inline docs/source/strategies/collaborative_training_expert.rst:11]:
```suggestion
Below are some ways to reduce communication when training collaboratively. As the size of your model increase, bottlenecks in communication become more apparent.
```

File: docs/source/strategies/collaborative_training_expert.rst [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+Compress Gradients & State
+""""""""""""""""""""""""""
+
+Hivemind allows you to compress gradients and states before sending them to other machines. This helps reduce the communication overhead substantially when training across the internet.
+
+Below we enable Float16 compression, which compresses gradients and states to Float16 before sending it to other machines.
```
Comment (RobertLaurella) [inline docs/source/strategies/collaborative_training_expert.rst:18]:
```suggestion
Below, we enable Float16 compression, which compresses gradients and states to Float16 before sending it to other machines.
```

File: docs/source/strategies/collaborative_training_expert.rst [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        strategy=CollaborativeStrategy(
+            target_batch_size=target_batch_size,
+            grad_compression=Float16Compression(),
+            state_averaging_compression=Float16Compression(),
+        ),
+        accelerator="gpu",
+        devices=1,
+    )
+
+A slightly more advanced scheme is dynamic compression based on the size of values. Below we enable 8-bit quantization for large numbers, and Float16 compression for small values.
```
Comment (RobertLaurella) [inline docs/source/strategies/collaborative_training_expert.rst:36]:
```suggestion
A slightly more advanced scheme is dynamic compression based on value size. Below, we enable 8-bit quantization for large numbers, and Float16 compression for small values.
```

File: docs/source/strategies/collaborative_training_intermediate.rst [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+Overlap Gradient and State Averaging
+""""""""""""""""""""""""""""""""""""
+
+When the target batch size is reached, all processes that are included in the step send gradients and model states to each other. By enabling some flags through
+the strategy, communication can happen in the background. This allows training to continue (with slightly outdated weights) but provides us the means
+to overlap communication with computation.
+
+.. warning::
+    Enabling these flags means that you must pass in a ``scheduler_fn`` to the ``CollaborativeStrategy`` instead of relying on a scheduler from ``configure_optimizers``.
+    The optimizer is re-created by Hivemind, and as a result, the scheduler has to as well.
```
Comment (RobertLaurella) [inline docs/source/strategies/collaborative_training_intermediate.rst:23]: The scheduler has to what - be re-created?

File: docs/source/strategies/collaborative_training_basic.rst (hunk 3 above)
Comment (RobertLaurella) [inline docs/source/strategies/collaborative_training_basic.rst:25]:
```suggestion
    When using ``CollaborativeStrategy`` note that you cannot use gradient accumulation (``accumulate_grad_batches``). This is because Hivemind manages accumulation internally.
```

File: docs/source/strategies/collaborative_training_intermediate.rst [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+Offloading Optimizer State to the CPU
+"""""""""""""""""""""""""""""""""""""
+
+Offloading the Optimizer state to the CPU works the same as :ref:`deepspeed-zero-stage-2-offload`, where we save GPU memory by keeping all optimizer states on the CPU.
+
+.. note::
+    We suggest enabling offloading and overlapping communication to hide the additional overhead from having to communicate with the CPU.
+
+.. warning::
```
[the remark this answers is not in the document]
Comment (RobertLaurella) [reply docs/source/strategies/collaborative_training_intermediate.rst:59]: +1

**Review by RobertLaurella — CHANGES_REQUESTED** [review]

File: docs/source/strategies/collaborative_training_expert.rst [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+Using Compression to Optimize communications
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Below are some ways to reduce communication when training collaboratively. As the model sizes get larger, bottlenecks in communication become more apparent.
+
+Compress Gradients & State
+""""""""""""""""""""""""""
+
+Hivemind allows you to compress gradients and states before sending them to other machines. This helps reduce the communication overhead substantially when training across the internet.
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_expert.rst:16]: I just added a note to mention this may affect convergence if going from Float32 -> Float16. 

I haven't done enough research around intra-node performance, but hivemind always uses the network adapter (local DDP support is a later thing to work on).

File: docs/source/strategies/collaborative_training_expert.rst [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+            grad_compression=Float16Compression(),
+            state_averaging_compression=Float16Compression(),
+        ),
+        accelerator="gpu",
+        devices=1,
+    )
+
+A slightly more advanced scheme is dynamic compression based on the size of values. Below we enable 8-bit quantization for large numbers, and Float16 compression for small values.
+
+Size Adaptive Compression has been used in a variety of Hivemind applications and has shown success.
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_expert.rst:38]: extended the note!

File: docs/source/strategies/collaborative_training_intermediate.rst [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+
+Re-using Gradient Buffers
+"""""""""""""""""""""""""
+
+By default, Hivemind accumulates gradients in a separate buffer. This means additional GPU memory is required to store gradients. You can disable this by passing ``reuse_grad_buffers=False`` to the ``CollaborativeStrategy``.
+
+.. warning::
+    The ``CollaborativeStrategy`` will override ``zero_grad`` in your ``LightningModule`` to have no effect. This is because gradients are accumulated in the model
+    and hivemind manages when they need to be cleared.
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_intermediate.rst]: I might be missing the point here, what are we customizing? The `zero_grad` function? We also throw a warning within the Strategy when the user has overridden this function and enabled `reuse_grad_buffers`.

File: docs/source/strategies/collaborative_training_intermediate.rst [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+        ),
+        accelerator="gpu",
+        devices=1,
+    )
+
+
+Re-using Gradient Buffers
+"""""""""""""""""""""""""
+
+By default, Hivemind accumulates gradients in a separate buffer. This means additional GPU memory is required to store gradients. You can disable this by passing ``reuse_grad_buffers=False`` to the ``CollaborativeStrategy``.
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_intermediate.rst:84]: Fixed the note!

File: docs/source/strategies/collaborative_training_expert.rst [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+    import pytorch_lightning as pl
+    from pytorch_lightning.strategies import CollaborativeStrategy
+    from functools import partial
+    from hivemind.optim.power_sgd_averager import PowerSGDGradientAverager
+
+    trainer = pl.Trainer(
+        strategy=CollaborativeStrategy(
+            target_batch_size=8192,
+            grad_averager_factory=partial(PowerSGDGradientAverager, averager_rank=32, min_compression_ratio=0.5),
```
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_expert.rst:83]: That's a great question, I was given these values but I wonder if there are ways to pick them other than trying different values.

cc @justheuristic

File: docs/source/strategies/collaborative_training_intermediate.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_intermediate.rst]: IF there is still confusion, will address in a future PR :)

File: docs/source/strategies/collaborative_training_expert.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_expert.rst:83]: Will address in a future PR, we might be able to automate some of these variables.

File: docs/source/strategies/collaborative_training_expert.rst [hunk 16]
```diff
@@ -0,0 +1,87 @@
+:orphan:
+
+.. _collaborative_training_expert:
+
+Training on unreliable mixed GPUs across the internet (Expert)
+==============================================================
+
+Using Compression to Optimize communications
```
Comment (akihironitta) [inline docs/source/strategies/collaborative_training_expert.rst:8]: nit 🐛 
```suggestion
Using Compression to Optimize Communications
```

File: docs/source/strategies/collaborative_training_intermediate.rst (hunk 2 above)
Comment (akihironitta) [reply docs/source/strategies/collaborative_training_intermediate.rst:95]: Can't see it used in the example below, but should it still stay here?

**Review by akihironitta — APPROVED** [review]
Haven't tried myself, but very interesting feature 🚀

File: docs/source/strategies/collaborative_training_intermediate.rst (hunk 2 above)
Comment (SeanNaren) [reply docs/source/strategies/collaborative_training_intermediate.rst:95]: eek read it wrong! removing

Comment (SeanNaren) [inline docs/source/strategies/collaborative_training_intermediate.rst:95]:
```suggestion
```

**Review by SeanNaren — COMMENTED** [review]
gatekeeper

**Review by RobertLaurella — APPROVED** [review]

**Review by Borda — APPROVED** [review]
